app/vmselect/netstorage/netstorage.go

```diff
@@ -3265,19 +3300,24 @@ func (pnc *perNodeCounter) GetTotal() uint64 {
 const maxFastAllocBlockSize = 32 * 1024
 
 // execSearchQuery calls cb for with marshaled requestData for each tenant in sq.
-func execSearchQuery(qt *querytracer.Tracer, sq *storage.SearchQuery, cb func(qt *querytracer.Tracer, requestData []byte, t storage.TenantToken) any) []any {
+func execSearchQuery(qt *querytracer.Tracer, sq *storage.SearchQuery, cancelled *atomic.Bool, cb func(qt *querytracer.Tracer, requestData []byte, t storage.TenantToken) any) []any {
 	var requestData []byte
 	var results []any
 
 	for i := range sq.TenantTokens {
+		// Stop processing the remaining tenants if the query is cancelled.
+		// It is safe to return the partial results.
+		if cancelled.Load() {
+			return results
+		}
 		requestData = sq.TenantTokens[i].Marshal(requestData)
 		requestData = sq.MarshaWithoutTenant(requestData)
 		qtL := qt
 		if sq.IsMultiTenant && qt.Enabled() {
 			qtL = qt.NewChild("query for tenant: %s", sq.TenantTokens[i].String())
```

Comment:
If panic happens at `qt.NewChild` - it requires to obtain a mutex protection before calling it. Because check above for `cancelled.Load()` is inefficient. This check may succeed, by trace may be closed at any moment after.

Reply:
With this PR `snr.finishQueryTracers` [waits](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/8124/files#diff-22edd94c512ef8044f9b703e982cb6e74928f8e4479afb5dcc7c1c16a2b7884bR2019-R2024) until all workers will be completed before closing `qt`.
This cancellation is intended to skip unnecessary work in order to reduce load on storage nodes when it is not needed.

Reply:
Looks hacky. And it may increase latency for all queries (including non-multitenant). Since vmselect now have to wait for the response of the slowest storage node.

 Also, `cancelled` could work only with `execSearchQuery`, it doenst make any sense for other callbacks, since it's executed immediately.

Reply:
> Looks hacky. And it may increase latency for all queries (including non-multitenant). Since vmselect now have to wait for the response of the slowest storage node.

Maybe you have another idea/approach to fixing this one which would work better?

> Also, cancelled could work only with execSearchQuery, it doenst make any sense for other callbacks, since it's executed immediately.

I'll remove this.

Reply:
I actually see only 1 way to fix this issue without latency increase - remove `finishQueryTracers` function. It was added as work around. There was a requirement to call Done(), now it's only print error log message.

 This solution do not fix potential data race, that could happen due to concurrent nature of storage requests. It should addressed in another pull request.

 The data race could happen, if collectResults function returned without waiting for the response of the remaining storages. In this case, during `Done` call content of the child tracers can be updated during read operations. It could case read write data race.

Reply:
Also, I think request `cancel` should be a part of the separate PR as a feature.